autoretrieve/engine.go

```diff
@@ -0,0 +1,524 @@
+// Engine internally uses "go-legs", a protocol for propagating and synchronizing changes an IPLD DAG, to publish advertisements.
+// See:
+//  - https://github.com/filecoin-project/storetheindex
+//  - https://github.com/filecoin-project/go-legs
+//
+// Published advertisements are signed using the given private key.
+// The retAddrs corresponds to the endpoints at which the data block associated to the advertised
+// multihashes can be retrieved.
+// Note that if no retAddrs is specified the listen addresses of the given libp2p host are used.
+//
+// The engine also provides the ability to generate advertisements via Engine.NotifyPut and
+// Engine.NotifyRemove as long as a provider.MultihashLister is registered.
+// See: provider.MultihashLister, Engine.RegisterMultihashLister.
+//
+// The engine must be started via Engine.Start before use and discarded via Engine.Shutdown when no longer needed.
+// See: Engine.Start, Engine.Shutdown.
+func New(o ...Option) (*AutoretrieveEngine, error) {
+	opts, err := newOptions(o...)
+	if err != nil {
+		return nil, err
+	}
+
+	e := &AutoretrieveEngine{
+		options: opts,
+	}
+
+	e.lsys = e.mkLinkSystem()
+
+	return e, nil
+}
+
+// Start starts the engine by instantiating the internal storage and joins the configured gossipsub
+// topic used for publishing advertisements.
+//
+// The context is used to instantiate the internal LRU cache storage.
+//
+// See: AutoretrieveEngine.Shutdown, chunker.NewCachedEntriesChunker, dtsync.NewPublisherFromExisting.
+func (e *AutoretrieveEngine) Start(ctx context.Context) error {
+	// Create datastore entriesChunker
+	entriesCacheDs := dsn.Wrap(e.ds, datastore.NewKey(linksCachePath))
+	cachedChunker, err := chunker.NewCachedEntriesChunker(ctx, entriesCacheDs, e.entChunkSize, e.entCacheCap)
+	if err != nil {
+		return err
+	}
+
+	if e.purgeCache {
+		err := cachedChunker.Clear(ctx)
+		if err != nil {
+			return err
+		}
+	}
+
+	e.entriesChunker = cachedChunker
+
+	e.publisher, err = e.newPublisher()
+	if err != nil {
+		log.Errorw("Failed to instantiate legs publisher", "err", err, "kind", e.pubKind)
+		return err
+	}
+
+	// Initialize publisher with latest advertisement CID.
+	adCid, err := e.getLatestAdCid(ctx)
+	if err != nil {
+		return fmt.Errorf("could not get latest advertisement cid: %w", err)
+	}
+	if adCid != cid.Undef {
+		if err = e.publisher.SetRoot(ctx, adCid); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *AutoretrieveEngine) newPublisher() (legs.Publisher, error) {
+	switch e.pubKind {
+	case NoPublisher:
+		log.Info("Remote announcements is disabled; all advertisements will only be store locally.")
+		return nil, nil
+	case DataTransferPublisher:
+		dtOpts := []dtsync.Option{dtsync.Topic(e.pubTopic), dtsync.WithExtraData(e.pubExtraGossipData)}
+		if e.pubDT != nil {
+			return dtsync.NewPublisherFromExisting(e.pubDT, e.h, e.pubTopicName, e.lsys, dtOpts...)
+		}
+		ds := dsn.Wrap(e.ds, datastore.NewKey("/legs/dtsync/pub"))
```

Comment:
why not make `/legs/dtsync/pub` as const like the others

Reply:
Same as above, https://github.com/filecoin-project/index-provider/

Reply:
again, it is a reference repo, we can improve things in our implementation